### PR TITLE
js-api: Add implementation defined limit of 16GiB for 64-bit memories.

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

**Comment by programmerjake** [comment]
why limit it to 16GiB? pretty reasonable programs can use waay more memory than that...

**Review by backes — APPROVED** [review]
LGTM, thanks!

**Comment by backes** [comment]
> why limit it to 16GiB? pretty reasonable programs can use waay more memory than that...

We need *some* spec'ed limit for Web embeddings to ensure predictability on the Web. We can always decide to raise it later, when the need arises. Any allocation above 16GiB is very unlikely to succeed on many clients anyway, and 16GiB is also the limit that the V8 sandbox currently imposes in Chrome.

File: document/js-api/index.bs [hunk 1]
```diff
@@ -1781,7 +1781,8 @@ In practice, an implementation may run out of resources for valid modules below
 
 <ul>
 <li>The maximum size of a table is 10,000,000.</li>
-<li>The maximum number of pages of a memory is 65,536.</li>
+<li>The maximum size of a 32-bit memory is 65,536 pages (4 GiB).</li>
```
Comment (dschuff) [inline document/js-api/index.bs:1784]: Is this actually necessary, given that more than 4G isn't even addressable?

Comment (sbc100) [reply document/js-api/index.bs:1784]: Do you think we should just drop this line then?

Comment (sbc100) [reply document/js-api/index.bs:1784]: I think its somewhat useful to have the two limits listed next to each other like this.. even though I guess its redundant.

**Comment by bvisness** [comment]
@sbc100 The implementation limit in Firefox currently stands at 8GiB, not 16. I think we'd be open to raising that limit, but the relevant people are on PTO right now so we'll have to get back to you later. In the meantime, we'd prefer that the spec say 8GiB.

**Comment by backes** [comment]
Oh, from https://searchfox.org/mozilla-central/source/js/src/wasm/WasmConstants.h#1156 we concluded that Firefox doesn't have any limit currently. Where is the 8GiB limit implemented?

**Comment by bvisness** [comment]
Ah that makes sense, sorry about the confusion. That constant is just for validation, not implementation. Here's the implementation limit:

https://searchfox.org/mozilla-central/source/js/src/wasm/WasmMemory.cpp#253
https://searchfox.org/mozilla-central/source/js/src/vm/ArrayBufferObject.h#218

File: document/js-api/index.bs (hunk 1 above)
Comment (dschuff) [reply document/js-api/index.bs:1784]: Oh nevermind, the limit itself is pre-existing.

**Comment by backes** [comment]
Thanks for clarifying, that makes sense!
Wouldn't it be enough to lower the validation limit to 16GiB then? It's always allowed to fail at runtime already before that limit.

We heard from partners that an 8GiB limit wouldn't be hugely helpful though given that a wasm64 program by design always needs more memory than the corresponding wasm32 program. So the 8GiB won't give you substantially more memory than the 4GiB for wasm32.
I would thus propose to work towards raising your implementation limit to 16GiB. We (in V8) are also open to raise the limit further once the need arises.

**Comment by eqrion** [comment]
> We heard from partners that an 8GiB limit wouldn't be hugely helpful though given that a wasm64 program by design always needs more memory than the corresponding wasm32 program. So the 8GiB won't give you substantially more memory than the 4GiB for wasm32. I would thus propose to work towards raising your implementation limit to 16GiB. We (in V8) are also open to raise the limit further once the need arises.

We'll look into it on our side. I don't know if there's anything fundamental about it, or it's just conservative. Some relevant folks are on PTO right now, so it'll take a while. If there is something fundamental about it, we may ask that the implementation limit only grow to 8 instead of 16.

**Comment by eqrion** [comment]
We think we'll be able to increase the limit to 16GiB. We're going to do the work here [1].

[1] https://bugzilla.mozilla.org/show_bug.cgi?id=1931401

**Pull request merged**